### ISSUE TITLE
Immediate crash after running native image for WAR [overlay]

### ISSUE THREAD

**Issue by mmoayyed** [spring-projects-experimental/spring-native#458]
I am using [this plugin](https://github.com/ayltai/spring-graalvm-native-plugin) to build a GraalVM native image. The configuration of the build is as follows:

- GraalVM Enterprise 21
- Repository: https://github.com/apereo/cas-overlay-template/tree/graal
- Gradle 6.8.1
- spring-graalvm-native-plugin 1.4.2 which pulls in 0.8.5

The build.gradle file contains the following blob:

```groovy
nativeImage {
    mainClassName = "org.apereo.cas.web.CasWebApplication"
    toolVersion = "21.0.0"
    springNativeMode = "reflection"
    reportExceptionStackTraces = true
    removeUnusedAutoConfig = true
    disableToolchainChecking = true
    removeSaturatedTypeFlows = true
    printAnalysisCallTree = true
    enableAllSecurityServices = true
    removeJmxSupport = true
    enableHttp = true
    enableHttps = true
    maxHeapSize = "8G"
    download = "skip"
    javaVersion = project.targetCompatibility
    initializeAtBuildTime = ["org.apache.commons.logging.LogFactory"] // should be removed in 0.9.0
}
```

To build a native image, one can use: `./gradlew buildNativeImage`:

```
cas:77540]     (clinit):   2,096.67 ms,  3.32 GB
[cas:77540]   (typeflow):  44,399.01 ms,  3.32 GB
[cas:77540]    (objects):  59,025.71 ms,  3.32 GB
[cas:77540]   (features):  10,102.77 ms,  3.32 GB
[cas:77540]     analysis: 119,557.48 ms,  3.32 GB
Printing call tree to /Users/Misagh/Workspace/GitWorkspace/cas-overlay-template/build/native/reports/call_tree_cas_20210129_183016.txt
Printing list of used methods to /Users/Misagh/Workspace/GitWorkspace/cas-overlay-template/build/native/reports/used_methods_cas_20210129_183046.txt
Printing list of used classes to /Users/Misagh/Workspace/GitWorkspace/cas-overlay-template/build/native/reports/used_classes_cas_20210129_183046.txt
Printing list of used packages to /Users/Misagh/Workspace/GitWorkspace/cas-overlay-template/build/native/reports/used_packages_cas_20210129_183047.txt
[cas:77540]     universe:   6,469.09 ms,  4.38 GB
[cas:77540]      (parse):  23,297.95 ms,  5.34 GB
[cas:77540]     (inline):  16,220.48 ms,  6.00 GB
[cas:77540]    (compile): 198,073.39 ms,  6.63 GB
[cas:77540]      compile: 250,935.04 ms,  6.63 GB
[cas:77540]        image:  18,000.52 ms,  6.38 GB
[cas:77540]        write:   4,333.42 ms,  6.38 GB
[cas:77540]      [total]: 580,916.84 ms,  6.38 GB

Deprecated Gradle features were used in this build, making it incompatible with Gradle 7.0.
Use '--warning-mode all' to show the individual deprecation warnings.
See https://docs.gradle.org/6.8.1/userguide/command_line_interface.html#sec:command_line_warnings

BUILD SUCCESSFUL in 9m 59s
6 actionable tasks: 4 executed, 2 up-to-date
````

This produces a `cas` native image:

```
lx build/native       
                                                                                                                                                              
Permissions Size User   Group Date Modified Name
drwxr-xr-x     - Misagh staff 29 Jan 18:25  BOOT-INF/
.rwxr-xr-x  125M Misagh staff 29 Jan 18:35  cas*
drwxr-xr-x     - Misagh staff 29 Jan 18:25  META-INF/
drwxr-xr-x     - Misagh staff 29 Jan 18:25  org/
drwxr-xr-x     - Misagh staff 29 Jan 18:30  reports/
```

My expectation is that I should be able to run the native image using `build/native/cas`. Unfortunately, all I see when I do is:

```
  .   ____          _            __ _ _
 /\\ / ___'_ __ _ _(_)_ __  __ _ \ \ \ \
( ( )\___ | '_ | '_| | '_ \/ _` | \ \ \ \
 \\/  ___)| |_)| | | | | || (_| |  ) ) ) )
  '  |____| .__|_| |_|_| |_\__, | / / / /
 =========|_|==============|___/=/_/_/_/
 :: Spring Boot ::                (v2.4.2)
```

Seems like the app crashes immediately. Could you recommend a way to troubleshoot the scenario here? The application has a custom banner; not sure where the SB banner itself comes from!

A few extra data points that might help:

1. The build is generally set to produce a `cas.war` file. The outcome is a WAR, not a JAR.
2. `java -jar build/libs/cas.war` works fine.
3. The gradle build is structured as a WAR overlay. Not sure if this creates complications.

**Comment by aclement** [comment]
My first suspicion would be that the jar containing CasWebApplication was not built using:
```
                <dependency>
                        <groupId>org.springframework</groupId>
                        <artifactId>spring-context-indexer</artifactId>
                </dependency>
```
The synthesis capability in the feature will not work if the components are in jars rather than in the application being built, as we can see from the output log for the build:

```
Found no META-INF/spring.components -> synthesizing one...
Computed spring.components is
vvv
^^^
```
Anything with spring components in should be built with that dependency.

For a custom banner we need to make sure the resource-config.json for the project includes it so that it is included in the resultant image, with something like
`src/main/resources/META-INF/resource-config.json`
```
{
"resources:{ "includes":[ {"pattern":"path/to/banner.txt"]}
}
```

**Comment by mmoayyed** [comment]
Thank you Andy for the follow-up. If I may,

- It's true that `spring-context-indexer` was not used when the jar containing CasWebApplication was built. I vaguely recall a few issues with indexer not quite allowing Spring to detect JPA entity classes that may be buried in some jar, but that was a long time ago and could have been a total miss on my part. I will give this another shot in the next few days.

- The `CasWebApplication` uses a custom banner that actually is a Java implementation of the Spring Boot banner abstraction, and it is buried in some jar somewhere. I trust this should not require extra config in a JSON file somewhere?

**Comment by mmoayyed** [comment]
Following up, I included the the spring-context-indexer and rebuilt the entire application stack. I can see now that `spring.components` files are generated for relevant JARs, which I proceeded to install into my local maven repository as well.


Next, I switched over to the WAR overlay variant (`graal` branch in the forementioned repo) and I modified the WAR overlay to point at the newly-installed artifacts (mainly switched versions to point at SNAPSHOT versions) and ran the build as always with `./gradlew buildNativeImage`. 

This is what I see in the outcome:

```
> Task :buildNativeImage
[cas:99599]    classlist:   8,325.02 ms,  1.71 GB
   _____                     _                             _   __           __     _              
  / ___/    ____    _____   (_)   ____    ____ _          / | / /  ____ _  / /_   (_) _   __  ___ 
  \__ \    / __ \  / ___/  / /   / __ \  / __ `/         /  |/ /  / __ `/ / __/  / / | | / / / _ \
 ___/ /   / /_/ / / /     / /   / / / / / /_/ /         / /|  /  / /_/ / / /_   / /  | |/ / /  __/
/____/   / .___/ /_/     /_/   /_/ /_/  \__, /         /_/ |_/   \__,_/  \__/  /_/   |___/  \___/ 
        /_/                            /____/                                                     
Feature operating in REFLECTION mode
Removing unused configurations
Removing JMX support
Use -Dspring.native.verbose=true on native-image call to see more detailed information from the feature
[cas:99599]        (cap):   2,572.12 ms,  1.71 GB
feature operating mode: reflection (spring init active? false)
Found #18 types in static reflection list to register
Skipping #16 types not on the classpath
Attempting proxy registration of #12 proxies
Skipped registration of #6 proxies - relevant types not on classpath
[cas:99599]        setup:   7,127.11 ms,  2.71 GB
Configuring initialization time for specific types and packages:
#139 buildtime-init-classes   #22 buildtime-init-packages   #18 runtime-init-classes    #0 runtime-init-packages

SLF4J: No SLF4J providers were found.
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#noProviders for further details.
Warning: class initialization of class org.springframework.boot.web.embedded.tomcat.TomcatServletWebServerFactory failed with exception java.lang.NoClassDefFoundError: org/apache/catalina/WebResourceRoot. This class will be initialized at run time because option --allow-incomplete-classpath is used for image building. Use the option --initialize-at-run-time=org.springframework.boot.web.embedded.tomcat.TomcatServletWebServerFactory to explicitly request delayed initialization of this class.
Warning: class initialization of class org.apache.commons.logging.LogAdapter$Log4jLog failed with exception java.lang.NoClassDefFoundError: org/apache/logging/log4j/LogManager. This class will be initialized at run time because option --allow-incomplete-classpath is used for image building. Use the option --initialize-at-run-time=org.apache.commons.logging.LogAdapter$Log4jLog to explicitly request delayed initialization of this class.
Feb 04, 2021 12:41:05 PM org.hibernate.validator.internal.util.Version <clinit>
INFO: HV000001: Hibernate Validator 6.1.7.Final
Registering resources - #0 bundles
Processing META-INF/spring.factories files...
Processing spring.factories - EnableAutoConfiguration lists #97 configurations
Excluding 53 auto-configurations from spring.factories file
Processing spring.factories - EnableAutoConfiguration lists #1 configurations
Excluding 0 auto-configurations from spring.factories file
Processing spring.factories - EnableAutoConfiguration lists #5 configurations
Excluding 0 auto-configurations from spring.factories file
Processing spring.factories - EnableAutoConfiguration lists #17 configurations
Excluding 6 auto-configurations from spring.factories file
Processing spring.factories - EnableAutoConfiguration lists #130 configurations
Reducing access on org.springframework.boot.autoconfigure.web.ServerProperties$Netty because HttpServer not around
Reducing access on org.springframework.boot.autoconfigure.web.ServerProperties$Undertow because Undertow not around
Reducing access on org.springframework.boot.autoconfigure.web.ServerProperties$Tomcat because Tomcat not around
Reducing access on org.springframework.boot.autoconfigure.web.ServerProperties$Jetty because WebAppContext not around
Reducing access on org.springframework.boot.autoconfigure.web.ServerProperties$Netty because HttpServer not around
Reducing access on org.springframework.boot.autoconfigure.web.ServerProperties$Undertow because Undertow not around
Reducing access on org.springframework.boot.autoconfigure.web.ServerProperties$Tomcat because Tomcat not around
Reducing access on org.springframework.boot.autoconfigure.web.ServerProperties$Jetty because WebAppContext not around
Reducing access on org.springframework.boot.autoconfigure.web.ServerProperties$Undertow because Undertow not around
Reducing access on org.springframework.boot.autoconfigure.web.ServerProperties$Netty because HttpServer not around
Reducing access on org.springframework.boot.autoconfigure.web.ServerProperties$Tomcat because Tomcat not around
Reducing access on org.springframework.boot.autoconfigure.web.ServerProperties$Jetty because WebAppContext not around
Reducing access on org.springframework.boot.autoconfigure.web.ServerProperties$Netty because HttpServer not around
Reducing access on org.springframework.boot.autoconfigure.web.ServerProperties$Undertow because Undertow not around
Reducing access on org.springframework.boot.autoconfigure.web.ServerProperties$Tomcat because Tomcat not around
Reducing access on org.springframework.boot.autoconfigure.web.ServerProperties$Jetty because WebAppContext not around
Excluding 94 auto-configurations from spring.factories file
spring.factories processing, problem adding access for key org.springframework.boot.liquibase.LiquibaseChangelogMissingFailureAnalyzer: liquibase/exception/ChangeLogParseException
Processing spring.factories - PropertySourceLoader lists #2 property source loaders
Processing existing META-INF/spring.components files...
Printing call tree to /Users/Misagh/Workspace/GitWorkspace/cas-overlay-template/build/native/reports/call_tree_cas_20210204_124106.txt
Printing list of used methods to /Users/Misagh/Workspace/GitWorkspace/cas-overlay-template/build/native/reports/used_methods_cas_20210204_124106.txt
Printing list of used classes to /Users/Misagh/Workspace/GitWorkspace/cas-overlay-template/build/native/reports/used_classes_cas_20210204_124106.txt
Printing list of used packages to /Users/Misagh/Workspace/GitWorkspace/cas-overlay-template/build/native/reports/used_packages_cas_20210204_124106.txt
Fatal error:java.lang.IllegalStateException: java.lang.IllegalStateException: No access hint found for import selector: org.springframework.cloud.client.discovery.EnableDiscoveryClientImportSelector
        at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
        at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
        at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:490)
        at java.base/java.util.concurrent.ForkJoinTask.getThrowableException(ForkJoinTask.java:600)
        at java.base/java.util.concurrent.ForkJoinTask.get(ForkJoinTask.java:1006)
        at com.oracle.svm.hosted.NativeImageGenerator.run(NativeImageGenerator.java:488)
        at com.oracle.svm.hosted.NativeImageGeneratorRunner.buildImage(NativeImageGeneratorRunner.java:370)
        at com.oracle.svm.hosted.NativeImageGeneratorRunner.build(NativeImageGeneratorRunner.java:529)
        at com.oracle.svm.hosted.NativeImageGeneratorRunner.main(NativeImageGeneratorRunner.java:119)
        at com.oracle.svm.hosted.NativeImageGeneratorRunner$JDK9Plus.main(NativeImageGeneratorRunner.java:561)
Caused by: java.lang.IllegalStateException: No access hint found for import selector: org.springframework.cloud.client.discovery.EnableDiscoveryClientImportSelector
        at org.springframework.nativex.type.Type.getHints(Type.java:1265)
        at org.springframework.nativex.support.ResourcesHandler.processType(ResourcesHandler.java:1249)
        at org.springframework.nativex.support.ResourcesHandler.processTypesToFollow(ResourcesHandler.java:1352)
        at org.springframework.nativex.support.ResourcesHandler.processType(ResourcesHandler.java:1295)
        at org.springframework.nativex.support.ResourcesHandler.processType(ResourcesHandler.java:960)
        at org.springframework.nativex.support.ResourcesHandler.checkAndRegisterConfigurationType(ResourcesHandler.java:950)
        at org.springframework.nativex.support.ResourcesHandler.processSpringComponent(ResourcesHandler.java:409)
        at org.springframework.nativex.support.ResourcesHandler.processSpringComponents(ResourcesHandler.java:371)
        at org.springframework.nativex.support.ResourcesHandler.handleSpringComponents(ResourcesHandler.java:250)
        at org.springframework.nativex.support.ResourcesHandler.register(ResourcesHandler.java:139)
        at org.springframework.nativex.support.SpringFeature.beforeAnalysis(SpringFeature.java:107)
        at com.oracle.svm.hosted.NativeImageGenerator.lambda$runPointsToAnalysis$7(NativeImageGenerator.java:701)
        at com.oracle.svm.hosted.FeatureHandler.forEachFeature(FeatureHandler.java:70)
        at com.oracle.svm.hosted.NativeImageGenerator.runPointsToAnalysis(NativeImageGenerator.java:701)
        at com.oracle.svm.hosted.NativeImageGenerator.doRun(NativeImageGenerator.java:563)
        at com.oracle.svm.hosted.NativeImageGenerator.lambda$run$0(NativeImageGenerator.java:476)
        at java.base/java.util.concurrent.ForkJoinTask$AdaptedRunnableAction.exec(ForkJoinTask.java:1407)
        at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
        at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
        at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
        at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
        at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:183)
Error: Image build request failed with exit status 1
```

Could you recommend a way to tackle the `No access hint found for import selector: org.springframework.cloud.client.discovery.EnableDiscoveryClientImportSelector` message? Is this something spring-native should have metadata for?

**Comment by aclement** [comment]
aha, some progress, cool! Ok so the problem with import selectors is that they don't typically statically declare what they might select so we can't work out which configurations they might return when called to select an import. For this reason we usually want hints for them. There is a line or two in the troubleshooting docs on this: https://repo.spring.io/milestone/org/springframework/experimental/spring-graalvm-native-docs/0.8.5/spring-graalvm-native-docs-0.8.5.zip!/reference/index.html#_no_access_hint_found_for_import_selector_xxx

you can either supply the hint it needs, or risk turning off the message (which may work if the selected import is one handled in another way, it might work).

If the app is complicated and contains libraries/elements we haven't tested, it could be worth just running it with the agent to collect all the config in one go. That is discussed here: https://repo.spring.io/milestone/org/springframework/experimental/spring-graalvm-native-docs/0.8.5/spring-graalvm-native-docs-0.8.5.zip!/reference/index.html#_missing_configuration - basically do one run in regular jvm mode with the agent and it will attempt to collect all needed config, than pass that to native-image.

**Comment by mmoayyed** [comment]
Thank you Andy. Following up on your first suggestion and feeling rather lucky :), I modified my Gradle plugin to enable warnings for missing selector hints. I am including a complete snippet for future passer-bys: 

```
nativeImage {
    mainClassName = "org.apereo.cas.web.CasWebApplication"
    toolVersion = "21.0.0"
    springNativeMode = "reflection"
    reportExceptionStackTraces = true
    removeUnusedAutoConfig = true
    disableToolchainChecking = true
    removeSaturatedTypeFlows = true
    printAnalysisCallTree = true
    enableAllSecurityServices = true
    removeJmxSupport = true
    warnMissingSelectorHints = true // this is the modified setting
    enableHttp = true
    enableHttps = true
    maxHeapSize = "8G"
    download = "skip"
    javaVersion = project.targetCompatibility
    initializeAtBuildTime = ["org.apache.commons.logging.LogFactory"]
}
```

With this, I was able to complete the build. Now, when running the native image, I see the following:

```
./build/native/cas                                                                                                                                                                                                                                                      

  .   ____          _            __ _ _
 /\\ / ___'_ __ _ _(_)_ __  __ _ \ \ \ \
( ( )\___ | '_ | '_| | '_ \/ _` | \ \ \ \
 \\/  ___)| |_)| | | | | || (_| |  ) ) ) )
  '  |____| .__|_| |_|_| |_\__, | / / / /
 =========|_|==============|___/=/_/_/_/
 :: Spring Boot ::                (v2.4.3)

Exception in thread "main" java.lang.IllegalStateException: The following classes could not be excluded because they are not auto-configuration classes:
	- org.springframework.boot.autoconfigure.orm.jpa.HibernateJpaAutoConfiguration
	- org.springframework.boot.autoconfigure.gson.GsonAutoConfiguration
	- org.springframework.boot.autoconfigure.data.redis.RedisAutoConfiguration

	at org.springframework.boot.autoconfigure.AutoConfigurationImportSelector.handleInvalidExcludes(AutoConfigurationImportSelector.java:216)
	at org.springframework.boot.autoconfigure.AutoConfigurationImportSelector.checkExcludedClasses(AutoConfigurationImportSelector.java:202)
	at org.springframework.boot.autoconfigure.AutoConfigurationImportSelector.getAutoConfigurationEntry(AutoConfigurationImportSelector.java:126)
	at org.springframework.boot.autoconfigure.AutoConfigurationImportSelector$AutoConfigurationGroup.process(AutoConfigurationImportSelector.java:434)
	at org.springframework.context.annotation.ConfigurationClassParser$DeferredImportSelectorGrouping.getImports(ConfigurationClassParser.java:879)
	at org.springframework.context.annotation.ConfigurationClassParser$DeferredImportSelectorGroupingHandler.processGroupImports(ConfigurationClassParser.java:809)
	at org.springframework.context.annotation.ConfigurationClassParser$DeferredImportSelectorHandler.process(ConfigurationClassParser.java:780)
	at org.springframework.context.annotation.ConfigurationClassParser.parse(ConfigurationClassParser.java:193)
	at org.springframework.context.annotation.ConfigurationClassPostProcessor.processConfigBeanDefinitions(ConfigurationClassPostProcessor.java:330)
	at org.springframework.context.annotation.ConfigurationClassPostProcessor.postProcessBeanDefinitionRegistry(ConfigurationClassPostProcessor.java:246)
```

...which probably is the result of (or related to) me ignoring warnings for selectors? I find it interesting/strange that the message reports on classes not being "auto-configuration classes" and yet clearly they are, judging by their name.

I'll try to collect the config first as you noted and then give it another try. 

As always, thanks much for the hints.

**Comment by aclement** [comment]
I'm not sure it is that problem actually. The clue is, as you suggest, "because they are not auto-configuration classes" - that suggests to me that something is excluding auto configs manually but they've already been thrown out for other reasons as the native-image was built. So its less about "they are not auto-configuration classes" and more accurately "I've no idea what they are because I can't find them" because we aren't making them accessible in the image as resources or via reflection.

Are you using `spring.autoconfigure.exclude` anywhere in a properties file in this application? If you are I'd say comment out that option for now, let native image do its work with throwing them out.

**Comment by mmoayyed** [comment]
Thanks. Your suggestion did set me on the right track. I realized the application was excluding a number of auto-configuration classes as part of its `SpringBootApplication` annotation in the main application class. So I removed those, built and installed the application again, and then switched over to the war-overlay project and restarted the native-image build. The errors went away and I was able to successfully complete the build. Then, when I ran the native-image, I got the Spring Boot stacktrace complaining about the missing `ServletWebServerFactory`:

```
Exception in thread "main" org.springframework.context.ApplicationContextException: Unable to start web server; nested exception is org.springframework.context.ApplicationContextException: Unable to start ServletWebServerApplicationContext due to missing ServletWebServerFactory bean.
	at org.springframework.boot.web.servlet.context.ServletWebServerApplicationContext.onRefresh(ServletWebServerApplicationContext.java:162)
	at org.apereo.cas.web.CasWebApplicationContext.onRefresh(CasWebApplicationContext.java:40)
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:576)
	at org.springframework.boot.web.servlet.context.ServletWebServerApplicationContext.refresh(ServletWebServerApplicationContext.java:144)
	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:767)
	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:759)
	at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:426)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:326)
	at org.springframework.boot.builder.SpringApplicationBuilder.run(SpringApplicationBuilder.java:144)
	at org.apereo.cas.web.CasWebApplication.main(CasWebApplication.java:71)
Caused by: org.springframework.context.ApplicationContextException: Unable to start ServletWebServerApplicationContext due to missing ServletWebServerFactory bean.
	at org.springframework.boot.web.servlet.context.ServletWebServerApplicationContext.getWebServerFactory(ServletWebServerApplicationContext.java:209)
	at org.springframework.boot.web.servlet.context.ServletWebServerApplicationContext.createWebServer(ServletWebServerApplicationContext.java:179)
	at org.springframework.boot.web.servlet.context.ServletWebServerApplicationContext.onRefresh(ServletWebServerApplicationContext.java:159)
	... 9 more
```

...which seemed strange since the WAR certainly does ship with the Spring Boot dependency on Tomcat as part of its overlay configuration. Nevertheless, I proceeded to include the following as an explicit dependency:

```gradle
    implementation "org.springframework.boot:spring-boot-starter-tomcat:${project.springBootVersion}"
```

...which got me past that error, and I am beginning to troubleshoot other issues that are related to the removal of auto-configuration classes and excludes, such as those that pop up with `GroovyTemplateAutoConfiguration`, etc. 


By the way, while I was trying to figure out the reason behind the missing `ServletWebServerFactory`, I tried a number of options before adding the Spring Boot module for tomcat and in one of those attempts, I did receive this error:

```
Error: Classes that should be initialized at run time got initialized during image building:
 javax.el.ListELResolver was unintentionally initialized at build time. To see why javax.el.ListELResolver got initialized use --trace-class-initialization=javax.el.ListELResolver

com.oracle.svm.core.util.UserError$UserException: Classes that should be initialized at run time got initialized during image building:
 javax.el.ListELResolver was unintentionally initialized at build time. To see why javax.el.ListELResolver got initialized use --trace-class-initialization=javax.el.ListELResolver

        at com.oracle.svm.core.util.UserError.abort(UserError.java:68)
        at com.oracle.svm.hosted.classinitialization.ConfigurableClassInitialization.checkDelayedInitialization(ConfigurableClassInitialization.java:539)
        at com.oracle.svm.hosted.classinitialization.ClassInitializationFeature.duringAnalysis(ClassInitializationFeature.java:226)
        at com.oracle.svm.hosted.NativeImageGenerator.lambda$runPointsToAnalysis$8(NativeImageGenerator.java:740)
        at com.oracle.svm.hosted.FeatureHandler.forEachFeature(FeatureHandler.java:70)
        at com.oracle.svm.hosted.NativeImageGenerator.runPointsToAnalysis(NativeImageGenerator.java:740)
        at com.oracle.svm.hosted.NativeImageGenerator.doRun(NativeImageGenerator.java:563)
        at com.oracle.svm.hosted.NativeImageGenerator.lambda$run$0(NativeImageGenerator.java:476)
```

Is `javax.el.ListELResolver` something for spring-native to handle directly?

**Comment by aclement** [comment]
> Is javax.el.ListELResolver something for spring-native to handle directly?

Could be. I see we do the initialization of it for ConfigServerAutoConfiguration but perhaps that trigger is too restrictive and we need it in a more general context, perfectly possible. You can of course set that yourself to see if it helps you. Just create an InitializationHint on your app and set it.
```
@InitializationHint(types = { javax.el.ListELResolver.class }, initTime = InitializationTime.BUILD)
```

I'm afraid I don't know anything about war overlays so debugging the rest is quite a challenge.

**Comment by mmoayyed** [comment]
Thanks for the suggestion. Looks like `InitializationHint` is only available in 0.9. I'll wait for the release. 

Some additional work in this area, if it helps:

- Had to add the following in `reflect-config.json`:

```groovy
    {
        "name":"org.springframework.boot.web.servlet.context.AnnotationConfigServletWebServerApplicationContext",
        "allDeclaredConstructors":true
    },
    {
        "name":"org.quartz.impl.StdSchedulerFactory", /* build complained about the missing default constructor */
        "allDeclaredConstructors":true
    }
```

- Had to add the following dependencies:

```groovy
    implementation "org.springframework.boot:spring-boot-starter-actuator:${project.springBootVersion}"
    implementation "org.springframework.boot:spring-boot-starter-tomcat:${project.springBootVersion}"

    /* 
		Fixes(?) ClassNotFoundException: org.springframework.scheduling.quartz.ResourceLoaderClassLoadHelper
	*/
    implementation "org.springframework.boot:spring-boot-starter-quartz:${project.springBootVersion}"
```

Unfortunately, after much trial and error the native-image app continues to crash on startup:

```
  .   ____          _            __ _ _
 /\\ / ___'_ __ _ _(_)_ __  __ _ \ \ \ \
( ( )\___ | '_ | '_| | '_ \/ _` | \ \ \ \
 \\/  ___)| |_)| | | | | || (_| |  ) ) ) )
  '  |____| .__|_| |_|_| |_\__, | / / / /
 =========|_|==============|___/=/_/_/_/
 :: Spring Boot ::                (v2.4.3)

Mar 07, 2021 1:23:08 PM org.apache.coyote.AbstractProtocol init
INFO: Initializing ProtocolHandler ["http-nio-8080"]
Mar 07, 2021 1:23:08 PM org.apache.catalina.core.StandardService startInternal
INFO: Starting service [Tomcat]
Mar 07, 2021 1:23:08 PM org.apache.catalina.core.StandardEngine startInternal
INFO: Starting Servlet engine: [Apache Tomcat/9.0.41]
Mar 07, 2021 1:23:08 PM org.apache.catalina.core.ApplicationContext log
INFO: Initializing Spring embedded WebApplicationContext
Mar 07, 2021 1:23:08 PM io.micrometer.core.instrument.binder.jvm.JvmGcMetrics isManagementExtensionsPresent
WARNING: GC notifications will not be available because MemoryPoolMXBeans are not provided by the JVM
Mar 07, 2021 1:23:08 PM org.apache.catalina.core.StandardService stopInternal
INFO: Stopping service [Tomcat]
```

I'll dig deeper in the next few days. Thanks for all the help thus far.

**Comment by sdeleuze** [comment]
Closing this issue since we need something more specific and reproducible with a sample to do something.
